File: typescript/packages/core/src/core/github/paginate.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import type { GitHubTransport } from './client.ts'

export async function githubPages(
  transport: GitHubTransport,
  path: string,
  opts: {
    params?: Record<string, string>
    limit?: number
    key?: string
    include?: (row: Record<string, unknown>) => boolean
  } = {},
): Promise<Record<string, unknown>[]> {
  const limit = opts.limit ?? 30
  if (limit < 1) return []
  const rows: Record<string, unknown>[] = []
  let page = 1
  const size = Math.min(100, limit)
  while (rows.length < limit) {
    const data = await transport.get(path, {
      ...(opts.params ?? {}),
      per_page: String(size),
      page: String(page),
    })
    const payload =
      opts.key !== undefined && data !== null && typeof data === 'object'
        ? (data as Record<string, unknown>)[opts.key]
        : data
    const batch = Array.isArray(payload)
      ? payload.filter(
          (item): item is Record<string, unknown> => item !== null && typeof item === 'object',
        )
      : []
    rows.push(...batch.filter((row) => opts.include?.(row) ?? true))
    if (batch.length < size) break
    page += 1
  }
  return rows.slice(0, limit)
}

File: python/mirage/core/github/paginate.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from collections.abc import Callable
from typing import Any

from mirage.core.github.client import github_request
from mirage.core.github.config import GhConfig


async def github_pages(
    config: GhConfig,
    path: str,
    *,
    params: dict[str, str] | None = None,
    limit: int = 30,
    key: str | None = None,
    include: Callable[[dict[str, Any]], bool]
    | None = None
) -> list[dict[str, Any]]:
    """Fetch a bounded REST list using GitHub's page/per_page contract."""
    if limit < 1:
        return []
    rows: list[dict[str, Any]] = []
    page = 1
    size = min(100, limit)
    while len(rows) < limit:
        query = {**(params or {}), "per_page": str(size), "page": str(page)}
        data = await github_request(config.token,
                                    "GET",
                                    path,
                                    params=query,
                                    base_url=config.base_url)
        payload: Any = data.get(key,
                                []) if key and isinstance(data, dict) else data
        batch = payload if isinstance(payload, list) else []
        candidates = (item for item in batch if isinstance(item, dict))
        rows.extend(item for item in candidates
                    if include is None or include(item))
        if len(batch) < size:
            break
        page += 1
    return rows[:limit]
